src/Puzzle.js

```diff
@@ -0,0 +1,237 @@
+            for (let m = 0; m < updatedCells.length; m++) {
+                let updatedCell = inverted ? updatedCells[ (updatedCells.length - 1) - m] : updatedCells[m];
+                if (updatedCell.value !== 0) {
+                    // We derive the note from the row and column, but use different logic than the launchpad to avoid playing far too low or high.
+                    let note = 30 + ((8 - updatedCell.row) * 8) + updatedCell.col;
+                    let onMessage = {
+                        type: "noteOn",
+                        channel: 0,
+                        velocity: 64,
+                        note: note
+                    };
+                    let offMessage = {
+                        type: "noteOff",
+                        channel: 0,
+                        velocity: 0,
+                        note: note
+                    };
+                    let timeOn = m * betweenNotes;
+                    let timeOff = timeOn + noteLength;
+                    for (let o = 0; o < this.outputCallbacks.length; o++) {
+                        let outputCallback = this.outputCallbacks[o];
+                        
+                        // TODO: replace with bergson
+                        setTimeout(
+                            () => { outputCallback(onMessage, /Launchpad/, true); },
+                            timeOn
+                            );
+                            setTimeout(
+                            ()=>{ outputCallback(offMessage, /Launchpad/, true); },
+                            timeOff
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    registerOutput (outputCallback) {
+        this.outputCallbacks.push(outputCallback);
+    }
+
+    handleMidiInput (midiMessage) {
+        if (midiMessage.type === "noteOn") {
+            let col = (midiMessage.note % 10) - 1;
+            let row = 8 - Math.floor(midiMessage.note / 10);
+            this.pushFromSquare(row, col);
+        }
+    }
+
+    pushFromSquare(row, col) {
+        let cellValue = this.state.grid[row][col];
+        // No move is possible if this is the empty square (value of 0).
+        if (cellValue === 0) {
+            // TODO: Add a sound or flash when an invalid key is pressed.
+        }
+        else {
+            var thisRow = this.state.grid[row];
+            var emptyCellColumnIndex = thisRow.indexOf(0);
+            
+            // We can shift pieces horizontally if our row contains the empty square.
+            if (emptyCellColumnIndex !== -1) {
+                var rowShiftedState = this.state.grid.slice();
+
+                // Rearrange the row so that the clicked position is now the empty square.            
+                var shiftedRow = this.shiftArray(thisRow, emptyCellColumnIndex, col);
+                rowShiftedState[row] = shiftedRow;
+
+                this.setState({grid: rowShiftedState});
```

Comment:
I am thinking through how to best refactor this and could use a little guidance.  I am currently avoiding setting state if I can tell there are no changes, which involves inspecting the current state. I assume that this isn't safe for async updates, but want to understand how to safely/performantly indicate that no state changes have taken place.  Is it like the Change Applier, where I can safely set it to the current state and have it make the call, or do I need to do anything else?

Reply:
For example, should I just return `{}` if there are no changes?

Reply:
Very good question. This is a little unsatisfying as I wasn't able to find the following in the docs, but I did find it in the React 16 release notes:

"Calling setState with null no longer triggers an update. This allows you to decide in an updater function if you want to re-render."

https://reactjs.org/blog/2017/09/26/react-v16.0.html#breaking-changes

So, you should be able to return `null` from the updater function. Returning `{}` is valid semantically, as the state change is merged into the state -- it will effectively leave the state unchanged. But my reading is that it will cause a component update.

Reply:
It's maybe worth leaving some notes here on approaches for optimizing/minimizing updates. React provides a lifecycle function `shouldComponentUpdate`:

https://reactjs.org/docs/react-component.html#shouldcomponentupdate

Which you can use to implement logic to determine if the component needs to be updated by inspecting the props and state. For example if our Puzzile `grid` state value is immutable, we could implement an efficient object reference check and only update if the old state `grid` is a different object from the new state `grid`. So, if we were to return `{}` from the state updater, it would cause an update, but our  `shouldComponentUpdate` implementation could easily catch that the `grid` hasn't actually changed.

React also provides `PureComponent` which provides a default implementation of `shouldComponentUpdate` -- a shallow comparison of props and state.

https://reactjs.org/docs/react-api.html#reactpurecomponent

Reply:
Additionally, due to the use of the shadow DOM, a full update won't actually update the browser DOM if the HTML content hasn't changed. React will call the component `render()` functions and compare the results against its shadow DOM. Only updating the parts of the browser DOM that have changed. This enables the implementation of 'naive' or simple full updates without causing a full element regeneration in the browser DOM. Which is good for performance and for not disturbing assistive technologies which might not be happy about parts of the web page being replaced.

https://reactjs.org/docs/reconciliation.html

Reply:
Thanks, this is all very useful for my understanding.

In this specific case, I don't think I should use `shouldComponentUpdate`.  Rather than comparing a change that is progress (`nextState`), I want to ponder a potential change and only actually trigger `setState` at all if a change is allowed.  For example, if you press the cell that corresponds to the open position, nothing should happen.  Same if you press a cell that does not share a row/column with the open position.

I think I'll use the `null` return pattern you shared, that seems like the best fit.